File: src/shared/domain/connection.ts
import { z } from 'zod';
import { capsSchema } from '../caps';

export const connectionKindSchema = z.enum([
  'postgres',
  'mariadb',
  'mysql',
  'sqlite',
  'clickhouse',
  'mongodb',
  'redis',
  'kafka',
  'sqs',
  's3',
]); // all v1 kinds; every one has an adapter as of P36
export type ConnectionKind = z.infer<typeof connectionKindSchema>;

// The connection dialog's default port per kind (D27's "kind-driven default port", not a
// second hardcoded number per adapter). Kinds with no conventional default port are absent.
export const DEFAULT_PORT: Partial<Record<ConnectionKind, number>> = {
  postgres: 5432,
  mariadb: 3306,
  mysql: 3306,
  // P36 D10: the HTTP interface, not the native protocol's 9000 — the only port the app's driver
  // (@clickhouse/client, HTTP-only) can ever speak to.
  clickhouse: 8123,
  mongodb: 27017,
  redis: 6379,
  kafka: 9092,
};

export const connectionColorSchema = z.enum([
  'none',
  'red',
  'orange',
  'amber',
  'olive',
  'green',
  'teal',
  'cyan',
  'blue',
  'indigo',
  'violet',
  'magenta',
  'grey',
]); // D18; matches --kira-conn-* in tokens.css. 'none' is a real, stored value (the P16 design
// system's default — "no colour is the default, the rail slot stays reserved either way") rather
// than the field being nullable, so no DB/schema change is needed to add it.
export type ConnectionColor = z.infer<typeof connectionColorSchema>;

export const connectionModeSchema = z.enum(['fields', 'uri']);
export type ConnectionMode = z.infer<typeof connectionModeSchema>;

// The plain object shape, with no refinement — kept separate so both connectionInputSchema
// (which adds the fields/uri superRefine below) and connectionSummarySchema (which cannot
// .omit() from a refined schema) can each build off it independently.
const connectionFieldsSchema = z.object({
  name: z.string().trim().min(1).max(120),
  kind: connectionKindSchema,
  color: connectionColorSchema,
  mode: connectionModeSchema,
  readOnly: z.boolean(),
  host: z.string().trim().nullable(),
  port: z.number().int().min(1).max(65535).nullable(),
  database: z.string().nullable(),
  username: z.string().nullable(),
  password: z.string().nullable(), // present on the way IN only; never on the way OUT (D9)
  uri: z.string().nullable(),
  options: z.record(z.string(), z.unknown()),
  // P11: optional shell command run before connect (e.g. a port-forward). A first-class column
  // rather than an options_json key — options round-trips through the connection URI and the
  // Copy URI menu item, and a shell command must never be settable by pasting a URI.
  preconnect: z.string().trim().min(1).max(2000).nullable().default(null),
  // Misc-fixes: overrides P11/D5's settle-window auto-detection with an explicit per-connection
  // choice. false (default) = "run each time it tries to connect" — a fresh instance is spawned
  // on every connect attempt and its exit is never monitored, whether or not it happens to still
  // be alive at the settle window. true = "run once, and disconnect the db when it dies" — always
  // arm() once the adapter connects, regardless of what the settle-window race resolved to (a
  // no-op if the script already exited, since there's nothing left to monitor).
  preconnectSidecar: z.boolean().default(false),
});

// SQS and S3 have no host/port at all (P10's D8, P17's own D8/D9 mirror) — fields mode repurposes
// `database` for the AWS region and `username` for the named profile instead, per §5.1's "named
// AWS profile" wording.
export const AWS_STYLE_KINDS: ReadonlySet<ConnectionKind> = new Set(['sqs', 's3']);

// Kinds whose "connection" is a local file path, not a network endpoint (P35 D10/D11). Fields
// mode repurposes `database` for the absolute path; host/port/username/password are unused.
export const FILE_KINDS: ReadonlySet<ConnectionKind> = new Set(['sqlite']);

export const connectionInputSchema = connectionFieldsSchema.superRefine((input, ctx) => {
  if (input.mode === 'fields') {
    if (FILE_KINDS.has(input.kind)) {
      const path = input.database?.trim() ?? '';
      if (!path) {
        ctx.addIssue({
          code: 'custom',
          path: ['database'],
          message: 'A database file is required.',
        });
      } else if (!path.startsWith('/')) {
        ctx.addIssue({
          code: 'custom',
          path: ['database'],
          message: 'The database file must be an absolute path.',
        });
      }
    } else {
      if (!AWS_STYLE_KINDS.has(input.kind) && !input.host) {
        ctx.addIssue({ code: 'custom', path: ['host'], message: 'Host is required.' });
      }
      if (!AWS_STYLE_KINDS.has(input.kind) && !input.port) {
        ctx.addIssue({ code: 'custom', path: ['port'], message: 'Port is required.' });
      }
    }
  } else {
    if (!input.uri || input.uri.trim() === '') {
      ctx.addIssue({ code: 'custom', path: ['uri'], message: 'A connection URI is required.' });
    }
  }
});

export type ConnectionInput = z.infer<typeof connectionInputSchema>;

// What the renderer gets. Note the absence of `password` — this is D9 enforced by the type.
export const connectionSummarySchema = connectionFieldsSchema.omit({ password: true }).extend({
  id: z.string(),
  sortOrder: z.number(),
  createdAt: z.string(),
  updatedAt: z.string(),
});
export type ConnectionSummary = z.infer<typeof connectionSummarySchema>;

export const connectionStatusSchema = z.enum(['disconnected', 'connecting', 'connected', 'error']);
export type ConnectionStatus = z.infer<typeof connectionStatusSchema>;

export const connectionStateSchema = z.object({
  connectionId: z.string(),
  status: connectionStatusSchema,
  serverVersion: z.string().nullable(),
  error: z.string().nullable(),
  since: z.number(), // epoch ms
  // Non-null only while connected — the toolbar's projection menu (Step 9) reads this to
  // decide whether server-side projection actually applies.
  caps: capsSchema.nullable(),
});
export type ConnectionState = z.infer<typeof connectionStateSchema>;

File: apps/kira-studio/internal/storage/model/connection.go
package model

// ConnectionFields is the shape both ConnectionsRepo.Insert/Update accept and ConnectionSummary
// embeds — packages/shared/domain/connection.ts's connectionFieldsSchema, minus `password` (D9: no
// password field anywhere in this package, enforced by the type itself, same as the TS build's
// `.omit({password: true})`).
type ConnectionFields struct {
	Name              string         `json:"name"`
	Kind              string         `json:"kind"`
	Color             string         `json:"color"`
	Mode              string         `json:"mode"`
	ReadOnly          bool           `json:"readOnly"`
	Host              *string        `json:"host"`
	Port              *int           `json:"port"`
	Database          *string        `json:"database"`
	Username          *string        `json:"username"`
	URI               *string        `json:"uri"`
	Options           map[string]any `json:"options"`
	Preconnect        *string        `json:"preconnect"`
	PreconnectSidecar bool           `json:"preconnectSidecar"`
	// P18 (v1.1) D18: runs the connection's own EXPLAIN before every SELECT a console run issues
	// on it, and warns when the threshold or a structural issue fires (never blocks). A first-class
	// column, not an options_json key — see the migration's own comment for why.
	AutoExplain bool `json:"autoExplain"`
	// P28 §5.3: commands per second this connection is paced to; 0 = unlimited (default). A
	// first-class column, not an options_json key — see the migration's own comment for why.
	ThrottlePerSec float64 `json:"throttlePerSec"`
}

// ConnectionSummary mirrors packages/shared/domain/connection.ts's connectionSummarySchema.
// ConnectionFields is embedded (not nested) so its fields JSON-inline into this struct's own
// object, matching the TS schema's flat shape.
type ConnectionSummary struct {
	ID string `json:"id"`
	ConnectionFields
	SortOrder int    `json:"sortOrder"`
	CreatedAt string `json:"createdAt"`
	UpdatedAt string `json:"updatedAt"`
}

// ConnectionState mirrors packages/shared/domain/connection.ts's connectionStateSchema. It is kept
// in-memory by the connections service (not read from a table) — listed here only as the shared
// wire type the repo's callers also use.
type ConnectionState struct {
	ConnectionID  string  `json:"connectionId"`
	Status        string  `json:"status"`
	ServerVersion *string `json:"serverVersion"`
	Error         *string `json:"error"`
	Since         int64   `json:"since"`
	Caps          any     `json:"caps"`
}

// connectionKinds mirrors connection.ts's connectionKindSchema — all v1 kinds.
var connectionKinds = map[string]bool{
	"postgres": true, "mariadb": true, "mysql": true, "sqlite": true, "clickhouse": true,
	"mongodb": true, "redis": true, "kafka": true, "sqs": true, "s3": true,
}

// paletteColors mirrors domain/color.ts's paletteColorSchema — the whole storable set, not
// PALETTE_COLOR_CHOICES/CONNECTION_COLOR_CHOICES (the picker's offered subset). P42 F27: a row
// saved with a retired colour must keep parsing, listing and painting its own rail, so trimming
// this set to match the picker would silently corrupt existing rows. P18 D18: renamed from
// connectionColors now that an environment's colour (this phase) draws from the same palette —
// ValidConnectionColor below stays as a one-line alias so repos/connections.go and the connection
// bridge are untouched.
var paletteColors = map[string]bool{
	"none": true, "red": true, "orange": true, "amber": true, "olive": true, "green": true,
	"teal": true, "cyan": true, "blue": true, "indigo": true, "violet": true, "magenta": true,
	"grey": true,
}

// ValidConnectionKind mirrors connection.ts's connectionKindSchema.
func ValidConnectionKind(v string) bool { return connectionKinds[v] }

// ValidPaletteColor mirrors domain/color.ts's paletteColorSchema (the full 13-colour set).
func ValidPaletteColor(v string) bool { return paletteColors[v] }

// ValidConnectionColor is ValidPaletteColor under its original name — kept so
// repos/connections.go and the connection bridge need no changes.
func ValidConnectionColor(v string) bool { return ValidPaletteColor(v) }

// ValidConnectionMode mirrors connection.ts's connectionModeSchema.
func ValidConnectionMode(v string) bool { return v == "fields" || v == "uri" }
